src/EventStore.js

```diff
@@ -75,11 +75,101 @@ class EventStore extends EventEmitter {
+        // would be doable relatively easily. Otherwise, we need to:
+        //  - store some other ordering information in the documents, e.g. monotonic timestamp
+        //  - define a deterministic order across partitions, e.g. by ordering by partitionId
+        //  - make sure the order at write-time holds the defined determinism, by guaranteeing that
+        //      given concurrent writes A (partitionId = 1, timestamp = X) and B (partitionId = 2, timestamp = X)
+        //      any index spanning both partitions will always have A come before B
+        //    If the index writer happens to receive B, it hence needs to make sure no write C (partitionId <= 2, timestamp <= X)
+        //    can still occur, before actually appending write B.
+        //    For all partitions part of the index, wait until it is guaranteed that no write to any of those will
+        //    happen with a timestamp <= current write timestamp. This can be guaranteed by delaying the index insert
+        //    by an amount of time that is the maximum delay for a write reaching the index writer. Since there is no guarantee
+        //    for an upper bound this needs to be tackled probabilistically, e.g. by measuring the 99th percentile of latencies.
+        //    Then on any index append, it needs to be checked that the determinism rule is still held. If not, an error needs
+        //    to be thrown, which would mean the according document write needs to be rolled back.
+        //    With this we basically have built a cross-partition transactional scheme, which is contrary to the goals.
+        //    So in contrast, no hard guarantee can be given for repeatable cross-partition reads.
+        this.storage.forEachPartition(partition => {
+            partition.readFi
+        });
+        const indexes = [];
+
+        let leastConsistentEntry = this.storage.index.lastEntry;
+        this.storage.forEachSecondaryIndex(index => {
+            indexes.push(index);
+            if (leastConsistentEntry.number > index.lastEntry.number) {
+                leastConsistentEntry = index.lastEntry;
+            }
+
+            const lastEntry = index.lastEntry;
+            // Start reading forward from that last index entry and update index to be in line with the storage again
+            let nextEvent, position = lastEntry.position;
+            while (nextEvent = this.storage.readFrom(lastEntry.partition, lastEntry.position, lastEntry.size)) {
+
+            }
+        });
+        let firstReindexSequenceNumber = leastConsistentEntry.number;
+        while (true) {
+            let document = this.storage.readFrom(partitionId, position)
+        }
+    }
+
+    /**
+     * Check if the last document in the storage is a torn write and if so, truncate the storage before that.
+     * @private
+     */
+    checkTornWrite() {
+        let position = this.storage.length;
```

Comment:
To fix torn writes (which should be a storage level concern) we need to infer the position from the actual documents in the storage rather than the information from the index. Hence the storage needs to read the last document of each partition and return the highest document header sequenceNumber.

Reply:
See #155